sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExecBase.scala

```diff
@@ -0,0 +1,239 @@
+    batchTimestampMs: Option[Long],
+    eventTimeWatermarkForEviction: Option[Long],
+    child: SparkPlan,
+    initialStateGroupingAttrs: Seq[Attribute],
+    initialState: SparkPlan)
+  extends BinaryExecNode
+  with StateStoreWriter
+  with WatermarkSupport
+  with TransformWithStateMetadataUtils {
+
+  override def operatorStateMetadataVersion: Int = 2
+
+  override def supportsSchemaEvolution: Boolean = true
+
+  override def left: SparkPlan = child
+
+  override def right: SparkPlan = initialState
+
+  // The keys that may have a watermark attribute.
+  override def keyExpressions: Seq[Attribute] = groupingAttributes
+
+  /**
+   * Distribute by grouping attributes - We need the underlying data and the initial state data to
+   * have the same grouping so that the data are co-located on the same task.
+   */
+  override def requiredChildDistribution: Seq[Distribution] = {
+    StatefulOperatorPartitioning.getCompatibleDistribution(
+      groupingAttributes,
+      getStateInfo,
+      conf) ::
+      StatefulOperatorPartitioning.getCompatibleDistribution(
+        initialStateGroupingAttrs,
+        getStateInfo,
+        conf) ::
+      Nil
+  }
+
+  /**
+   * We need the initial state to also use the ordering as the data so that we can co-locate the
+   * keys from the underlying data and the initial state.
+   */
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = Seq(
+    groupingAttributes.map(SortOrder(_, Ascending)),
+    initialStateGroupingAttrs.map(SortOrder(_, Ascending)))
+
+  override def shouldRunAnotherBatch(newInputWatermark: Long): Boolean = {
+    if (timeMode == ProcessingTime) {
+      // TODO SPARK-50180: check if we can return true only if actual timers are registered,
+      //  or there is expired state
+      true
+    } else if (outputMode == OutputMode.Append || outputMode == OutputMode.Update) {
+      eventTimeWatermarkForEviction.isDefined &&
+      newInputWatermark > eventTimeWatermarkForEviction.get
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Controls watermark propagation to downstream modes. If timeMode is ProcessingTime, the output
+   * rows cannot be interpreted in eventTime, hence this node will not propagate watermark in this
+   * timeMode.
+   *
+   * For timeMode EventTime, output watermark is same as input Watermark because
+   * transformWithState does not allow users to set the event time column to be earlier than the
+   * watermark.
+   */
+  override def produceOutputWatermark(inputWatermarkMs: Long): Option[Long] = {
+    timeMode match {
+      case ProcessingTime =>
+        None
+      case _ =>
+        Some(inputWatermarkMs)
+    }
+  }
+
+  // operator specific metrics
+  override def customStatefulOperatorMetrics: Seq[StatefulOperatorCustomMetric] = {
+    Seq(
+      // metrics around initial state
+      StatefulOperatorCustomSumMetric(
+        "initialStateProcessingTimeMs",
+        "Number of milliseconds taken to process all initial state"),
```

Comment:
@bogao007 @anishshri-db some metrics were missing in pyspark exec, but I think it's fine to have additional metrics.